File: common/src/main/java/xyz/qincai/manhunt/game/GameManager.java
package xyz.qincai.manhunt.game;

import xyz.qincai.manhunt.platform.ManhuntPlatform;
import xyz.qincai.manhunt.player.PlayerRole;

import java.util.UUID;
import java.util.concurrent.atomic.AtomicInteger;

public class GameManager {
    private final ManhuntPlatform platform;
    private final Match match;
    private int countdownTaskId = -1;
    private boolean forceStart;

    public GameManager(ManhuntPlatform platform) {
        this.platform = platform;
        this.match = new Match();
    }

    public Match getMatch() { return match; }

    public ManhuntPlatform getPlatform() { return platform; }

    private void resetMatchTiming() {
        match.setEndTime(0);
        match.setPausedAt(0);
        match.setTotalPausedDuration(0);
        match.setStartTime(0);
    }

    public void startGame() { startGame(null); }

    public void startGame(UUID ownerUuid) {
        if (match.getState() != GameState.WAITING) return;
        if (match.getRunnerUuid() == null) {
            platform.getUIFacade().broadcastMessage("\u00a7cNo runner selected!");
            return;
        }
        if (match.getHunterUuids().isEmpty()) {
            platform.getUIFacade().broadcastMessage("\u00a7cNo hunters selected!");
            return;
        }
        match.setOwnerUuid(ownerUuid);
        match.setState(GameState.COUNTDOWN);
        startCountdown();
    }

    public void startGameForce(UUID ownerUuid) {
        if (match.getState() != GameState.WAITING) return;
        match.setOwnerUuid(ownerUuid);
        forceStart = true;
        match.setState(GameState.COUNTDOWN);
        startCountdown();
    }

    private void startCountdown() {
        int initialCountdown = platform.getConfigProvider().getPreHuntCountdown();
        match.setState(GameState.COUNTDOWN);
        platform.getWorldProvider().freezeAllPlayers(match);
        AtomicInteger countdown = new AtomicInteger(initialCountdown);

        countdownTaskId = platform.getScheduler().runTaskTimer(() -> {
            if (match.getState() != GameState.COUNTDOWN) {
                platform.getScheduler().cancelTask(countdownTaskId);
                return;
            }
            int current = countdown.get();
            if (current <= 0) {
                platform.getScheduler().cancelTask(countdownTaskId);
                finishCountdown();
                return;
            }
            String title = "\u00a7e" + current;
            String subtitle = "\u00a77Game starting in...";
            for (UUID uuid : match.getHunterUuids()) {
                platform.getWorldProvider().sendTitle(uuid, title, subtitle);
            }
            if (match.getRunnerUuid() != null) {
                platform.getWorldProvider().sendTitle(match.getRunnerUuid(), title, subtitle);
            }
            countdown.decrementAndGet();
        }, 0L, 20L);
    }

    private void finishCountdown() {
        platform.getWorldProvider().createGameWorlds(match);
        String gameWorldName = match.getGameWorldName();
        if (gameWorldName == null) {
            platform.getUIFacade().broadcastMessage("\u00a7cFailed to create game world!");
            match.setState(GameState.WAITING);
            return;
        }
        platform.getWorldProvider().clearPlayerState(match);
        platform.getWorldProvider().teleportToFormation(match);
        platform.getWorldProvider().setAllPlayersSurvival(match);
        platform.getWorldProvider().healAllPlayers(match);

        match.setStrongholdDiscovered(false);
        match.setFortressDiscovered(false);
        match.setBlazeRodObtained(false);
        match.setBastionDiscovered(false);
        resetMatchTiming();

        if (forceStart) {
            forceStart = false;
            match.setState(GameState.RUNNING);
            match.setStartTime(System.currentTimeMillis());
            platform.getWorldProvider().unfreezeAllPlayers(match);
            if (match.getRunnerUuid() != null) {
                platform.getWorldProvider().setInvulnerable(match.getRunnerUuid(), false);
            }
            platform.getUIFacade().sendTitle("\u00a7cThe Hunt Has Begun!", "\u00a77Runner is on the loose!");
            platform.getUIFacade().sendToAll("\u00a7aThe hunt has started! (Force started)");
        } else {
            match.setState(GameState.PRE_HUNT);
            platform.getWorldProvider().unfreezeHorizontalAllPlayers(match);
            platform.getUIFacade().sendTitle("\u00a76Pre-Hunt", "\u00a77Damage a hunter to start the hunt");
            platform.getUIFacade().sendToAll("\u00a7ePre-Hunt phase! Runner must damage a hunter to start the hunt.");
        }
    }

    public void startHunt() {
        if (match.getState() != GameState.PRE_HUNT) return;
        match.setState(GameState.RUNNING);
        resetMatchTiming();
        match.setStartTime(System.currentTimeMillis());
        match.setStrongholdDiscovered(false);
        match.setFortressDiscovered(false);
        match.setBlazeRodObtained(false);
        match.setBastionDiscovered(false);
        platform.getWorldProvider().unfreezeAllPlayers(match);
        if (match.getRunnerUuid() != null) {
            platform.getWorldProvider().setInvulnerable(match.getRunnerUuid(), false);
        }
        platform.getUIFacade().sendTitle("\u00a7cThe Hunt Has Begun!", "\u00a77Runner is on the loose!");
        platform.getUIFacade().sendToAll("\u00a7aThe hunt has started!");
    }

    public void stopGame() {
        match.accumulatePausedTime();
        match.setState(GameState.FINISHED);
        match.setEndTime(System.currentTimeMillis());
        if (countdownTaskId != -1) {
            platform.getScheduler().cancelTask(countdownTaskId);
            countdownTaskId = -1;
        }
        platform.getUIFacade().stopUIUpdates();
        platform.getWorldProvider().clearEffects(match);
        platform.getWorldProvider().unfreezeAllPlayers(match);
        platform.getWorldProvider().teleportToMainWorld(match);
        match.setState(GameState.WAITING);
        match.setSeed(null);
        match.setWorldName(null);
        platform.getPlayerRegistry().reset();
    }

    public void runnerWins() {
        match.setState(GameState.FINISHED);
        match.setEndTime(System.currentTimeMillis());
        match.setRunnerWins(true);
        platform.getUIFacade().stopUIUpdates();
        platform.getWorldProvider().clearEffects(match);
        platform.getUIFacade().sendTitle("\u00a76Runner Wins!", "\u00a77The Ender Dragon has been defeated!");
        platform.getUIFacade().broadcastMessage("\u00a76\u00a7lRunner has won the game!");
        platform.getWorldProvider().unfreezeAllPlayers(match);
        platform.getWorldProvider().teleportToMainWorld(match);
        platform.getScheduler().runTaskTimer(() -> {
            match.setState(GameState.WAITING);
            match.setSeed(null);
            match.setWorldName(null);
            platform.getPlayerRegistry().reset();
        }, 200L, 0L);
        platform.getScheduler().cancelTask(countdownTaskId);
    }

    public void huntersWin() {
        match.setState(GameState.FINISHED);
        match.setEndTime(System.currentTimeMillis());
        match.setRunnerWins(false);
        platform.getUIFacade().stopUIUpdates();
        platform.getWorldProvider().clearEffects(match);
        platform.getUIFacade().sendTitle("\u00a7cHunters Win!", "\u00a77The Runner has been eliminated!");
        platform.getUIFacade().broadcastMessage("\u00a7c\u00a7lHunters have won the game!");
        platform.getWorldProvider().unfreezeAllPlayers(match);
        platform.getWorldProvider().teleportToMainWorld(match);
        platform.getScheduler().runTaskTimer(() -> {
            match.setState(GameState.WAITING);
            match.setSeed(null);
            match.setWorldName(null);
            platform.getPlayerRegistry().reset();
        }, 200L, 0L);
        platform.getScheduler().cancelTask(countdownTaskId);
    }

    public boolean isGameActive() {
        GameState s = match.getState();
        return s == GameState.RUNNING || s == GameState.PRE_HUNT || s == GameState.PAUSED;
    }

    public boolean pauseGame() {
        if (match.getState() != GameState.RUNNING && match.getState() != GameState.PRE_HUNT) return false;
        pauseInternal();
        return true;
    }

    public boolean pauseGame(UUID ownerUuid) {
        if (match.getState() != GameState.RUNNING && match.getState() != GameState.PRE_HUNT) return false;
        if (!match.isOwner(ownerUuid)) return false;
        pauseInternal();
        return true;
    }

    private void pauseInternal() {
        match.setPrePauseState(match.getState());
        match.setState(GameState.PAUSED);
        match.setPausedAt(System.currentTimeMillis());
        platform.getWorldProvider().freezeAllPlayers(match);
        platform.getWorldProvider().setGameRule(match, "doDaylightCycle", false);
        platform.getWorldProvider().setAllPlayersInvulnerable(match, true);
        platform.getWorldProvider().clearMobTargets(match);
        platform.getUIFacade().showPauseTitle();
        platform.getUIFacade().sendToAll("\u00a7eGame has been paused!");
    }

    public boolean resumeGame(UUID ownerUuid) {
        if (match.getState() != GameState.PAUSED) return false;
        if (!match.isOwner(ownerUuid)) return false;
        match.accumulatePausedTime();
        GameState previousState = match.getPrePauseState();
        match.setState(previousState);
        platform.getWorldProvider().setGameRule(match, "doDaylightCycle", true);
        platform.getWorldProvider().setAllPlayersInvulnerable(match, false);
        if (previousState == GameState.RUNNING) {
            platform.getWorldProvider().unfreezeAllPlayers(match);
        } else if (previousState == GameState.PRE_HUNT) {
            platform.getWorldProvider().unfreezeHorizontalAllPlayers(match);
        }
        platform.getUIFacade().hidePauseTitle();
        platform.getUIFacade().sendTitle("\u00a7aGame Resumed", "\u00a77The game has been resumed");
        platform.getUIFacade().sendToAll("\u00a7aGame has been resumed!");
        return true;
    }

    public boolean isGamePaused() { return match.getState() == GameState.PAUSED; }
}

File: pumpkin-plugin/game_manager.py
import time
from typing import Optional

from game_state import GameState, StartMode, ManhuntGameMode
from match import Match
from player_role import PlayerRole


class GameManager:
    def __init__(self):
        self.match = Match()
        self.force_start: bool = False
        self.headstart_remaining: int = -1
        self._countdown_active: bool = False
        self._headstart_active: bool = False

    def start_game(self, owner_uuid: Optional[str] = None) -> bool:
        if self.match.state != GameState.WAITING:
            return False

        if not self.match.runner_uuids:
            return False
        if not self.match.hunter_uuids:
            return False

        self.match.owner_uuid = owner_uuid
        self.match.state = GameState.COUNTDOWN
        return True

    def start_game_force(self, owner_uuid: Optional[str] = None) -> bool:
        if self.match.state != GameState.WAITING:
            return False

        self.match.owner_uuid = owner_uuid
        self.force_start = True
        self.match.state = GameState.COUNTDOWN
        return True

    def finish_countdown(self) -> None:
        self.match.stronghold_discovered = False
        self.match.fortress_discovered = False
        self.match.blaze_rod_obtained = False
        self.match.bastion_discovered = False

        self._reset_match_timing()

        if self.match.start_mode == StartMode.HEADSTART:
            self.match.state = GameState.HEADSTART
            self._headstart_active = True
        elif self.force_start:
            self.force_start = False
            self.match.state = GameState.RUNNING
            self.match.start_time = int(time.time() * 1000)
        else:
            self.match.state = GameState.PRE_HUNT

    def start_hunt(self) -> bool:
        if self.match.state != GameState.PRE_HUNT:
            return False

        self.match.state = GameState.RUNNING
        self._reset_match_timing()
        self.match.start_time = int(time.time() * 1000)
        self.match.stronghold_discovered = False
        self.match.fortress_discovered = False
        self.match.blaze_rod_obtained = False
        self.match.bastion_discovered = False
        return True

    def stop_game(self) -> None:
        self.match.accumulate_paused_time()
        self.match.state = GameState.FINISHED
        self.match.end_time = int(time.time() * 1000)
        self._countdown_active = False
        self._headstart_active = False

        self.match.state = GameState.WAITING
        self.match.seed = None
        self.match.world_name = None
        self.match.clear_all_players()

    def runner_wins(self) -> None:
        self.match.state = GameState.FINISHED
        self.match.end_time = int(time.time() * 1000)
        self.match.runner_wins = True

        self.match.state = GameState.WAITING
        self.match.seed = None
        self.match.world_name = None
        self.match.clear_all_players()

    def hunters_win(self) -> None:
        self.match.state = GameState.FINISHED
        self.match.end_time = int(time.time() * 1000)
        self.match.runner_wins = False

        self.match.state = GameState.WAITING
        self.match.seed = None
        self.match.world_name = None
        self.match.clear_all_players()

    def is_game_active(self) -> bool:
        return self.match.state in (
            GameState.RUNNING,
            GameState.PRE_HUNT,
            GameState.HEADSTART,
            GameState.PAUSED,
        )

    def is_game_paused(self) -> bool:
        return self.match.state == GameState.PAUSED

    def pause_game(self, owner_uuid: Optional[str] = None) -> bool:
        if self.match.state not in (
            GameState.RUNNING,
            GameState.PRE_HUNT,
            GameState.HEADSTART,
        ):
            return False

        if owner_uuid is not None and not self.match.is_owner(owner_uuid):
            return False

        self.match.pre_pause_state = self.match.state
        self.match.state = GameState.PAUSED
        self.match.paused_at = int(time.time() * 1000)
        self._headstart_active = False
        return True

    def resume_game(self, owner_uuid: Optional[str] = None) -> bool:
        if self.match.state != GameState.PAUSED:
            return False

        if owner_uuid is not None and not self.match.is_owner(owner_uuid):
            return False

        self.match.accumulate_paused_time()
        previous = self.match.pre_pause_state
        if previous is not None:
            self.match.state = previous
        return True

    def infect_player(self, runner_uuid: str) -> bool:
        if self.match.game_mode != ManhuntGameMode.INFECTION:
            return False
        if self.match.state != GameState.RUNNING:
            return False
        if runner_uuid not in self.match.runner_uuids:
            return False

        self.match.remove_runner(runner_uuid)
        self.match.add_hunter(runner_uuid)

        if not self.match.runner_uuids:
            self.hunters_win()
            return True
        return True

    def _reset_match_timing(self) -> None:
        self.match.end_time = 0
        self.match.paused_at = 0
        self.match.total_paused_duration = 0
        self.match.start_time = 0
        self.match.headstart_remaining = -1
